src/v/cluster/controller_snapshot.h

```diff
@@ -11,22 +11,136 @@
 
 #pragma once
 
+#include "cluster/config_manager.h"
 #include "cluster/types.h"
+#include "features/feature_table_snapshot.h"
 #include "serde/envelope.h"
 #include "serde/serde.h"
+#include "utils/fragmented_vector.h"
+
+#include <absl/container/btree_map.h>
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <absl/container/node_hash_map.h>
 
 namespace cluster {
 
+namespace controller_snapshot_parts {
+
+struct bootstrap_t
+  : public serde::
+      envelope<bootstrap_t, serde::version<0>, serde::compat_version<0>> {
+    std::optional<model::cluster_uuid> cluster_uuid;
+
+    friend bool operator==(const bootstrap_t&, const bootstrap_t&) = default;
+
+    auto serde_fields() { return std::tie(cluster_uuid); }
+};
+
+struct features_t
+  : public serde::
+      envelope<features_t, serde::version<0>, serde::compat_version<0>> {
+    features::feature_table_snapshot snap;
+
+    friend bool operator==(const features_t&, const features_t&) = default;
+
+    auto serde_fields() { return std::tie(snap); }
+};
+
+struct members_t
+  : public serde::
+      envelope<members_t, serde::version<0>, serde::compat_version<0>> {
+    struct node_t
+      : serde::envelope<node_t, serde::version<0>, serde::compat_version<0>> {
+        model::broker broker;
+        broker_state state;
+
+        friend bool operator==(const node_t&, const node_t&) = default;
+
+        auto serde_fields() { return std::tie(broker, state); }
+    };
+
+    struct update_t
+      : serde::envelope<update_t, serde::version<0>, serde::compat_version<0>> {
+        node_update_type type;
+        model::offset offset;
+        std::optional<model::revision_id> decommission_update_revision;
+
+        friend bool operator==(const update_t&, const update_t&) = default;
+
+        auto serde_fields() {
+            return std::tie(type, offset, decommission_update_revision);
+        }
+    };
+
+    absl::flat_hash_map<model::node_uuid, model::node_id> node_ids_by_uuid;
+    model::node_id next_assigned_id;
+
+    absl::node_hash_map<model::node_id, node_t> nodes;
+    absl::node_hash_map<model::node_id, node_t> removed_nodes;
+    absl::flat_hash_set<model::node_id> removed_nodes_still_in_raft0;
+    absl::node_hash_map<model::node_id, update_t> in_progress_updates;
+
+    model::offset first_node_operation_command_offset;
+
+    friend bool operator==(const members_t&, const members_t&) = default;
+
+    auto serde_fields() {
+        return std::tie(
+          node_ids_by_uuid,
+          next_assigned_id,
+          nodes,
+          removed_nodes,
+          removed_nodes_still_in_raft0,
+          in_progress_updates,
+          first_node_operation_command_offset);
+    }
+};
+
+struct config_t
+  : public serde::
+      envelope<config_t, serde::version<0>, serde::compat_version<0>> {
+    config_version version;
+    absl::btree_map<ss::sstring, ss::sstring> values;
```

Comment:
nit: in future we might want to revisit this and make the value something that has a timestamp, so that we don't lose the history of which property was set when